tachyon/zk/air/constraint_system/constraint_system.h

```diff
@@ -0,0 +1,121 @@
+      }
+    }
+    for (size_t row : {0, 1}) {
+      for (Eigen::Index col = 0; col < main_width_; ++col) {
+        main_.push_back(Variable::Main(row, col));
+      }
+    }
+    for (size_t i = 0; i < num_public_values; ++i) {
+      public_values_.push_back(Variable::Public(i));
+    }
+  }
+
+  const std::vector<Variable>& public_values() const { return public_values_; }
+
+  const std::vector<Variable>& preprocessed() const { return preprocessed_; }
+
+  const std::vector<Variable>& main() const { return main_; }
+
+  void EnforceFirstRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::FirstRow(std::move(constraint)));
+  }
+
+  void EnforceLastRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::LastRow(std::move(constraint)));
+  }
+
+  void EnforceTransitionConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::Transition(std::move(constraint)));
+  }
+
+  void EnforceConstraint(std::unique_ptr<Expression<F>> constraint) {
+    constraints_.push_back(std::move(constraint));
+  }
+
+  bool IsSatisfied(Evaluator<F>& evaluator, const std::vector<F>& public_values,
+                   const Eigen::Ref<const math::RowMajorMatrix<F>> main) const {
+    CHECK(public_values.size() == num_public_values_);
+    CHECK(main.cols() == main_width_);
+
+    for (Eigen::Index i = 0; i < main.rows() - 1; ++i) {
+      for (const auto& constraint : constraints_) {
+        if (!evaluator
+                 .Evaluate(constraint.get(), i, main.rows(), public_values,
+                           std::nullopt, main.middleRows(i, 2))
+                 .IsZero()) {
+          NOTREACHED() << i << constraint->ToString();
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  bool IsSatisfied(Evaluator<F>& evaluator, const std::vector<F>& public_values,
+                   const math::RowMajorMatrix<F>& preprocessed,
+                   const math::RowMajorMatrix<F>& main) const {
+    CHECK(public_values.size() == num_public_values_);
+    CHECK(main.cols() == main_width_);
+    CHECK(preprocessed.cols() == preprocessed_width_);
+    CHECK(preprocessed.rows() == main.rows());
+
+    for (Eigen::Index i = 0; i < main.rows() - 1; ++i) {
+      for (const auto& constraint : constraints_) {
+        if (!evaluator
+                 .Evaluate(constraint.get(), i, main.rows(), public_values,
+                           preprocessed.middleRows(i, 2), main.middleRows(i, 2))
+                 .IsZero()) {
+          NOTREACHED() << i << constraint->ToString();
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+ private:
+  Eigen::Index preprocessed_width_ = 0;
+  Eigen::Index main_width_ = 2;
+  size_t num_public_values_ = 0;
+  std::vector<Variable> preprocessed_;
+  std::vector<Variable> main_;
+  std::vector<Variable> public_values_;
```

Comment:
Why they are needed?

Reply:
Indeed they are not needed if we manually construct `Variable`when building the constraints. This is taken from [Plonky3 implementation.](https://github.com/Plonky3/Plonky3/blob/7c5deb0eab7191a97f7bb088637c1a68b2e6eb68/uni-stark/src/symbolic_builder.rs#L76). Do you want me to remove?

Reply:
Yes we don't need to follow all if the things are unneeded.